extensions/mds-logginghouse-binder/src/main/java/de/sovity/edc/extension/mdslogginhousebinder/MdsLoggingHouseBinder.java

```diff
@@ -0,0 +1,92 @@
+import de.sovity.edc.extension.contacttermination.ContractAgreementTerminationService;
+import de.sovity.edc.extension.contacttermination.ContractTerminationEvent;
+import de.sovity.edc.extension.contacttermination.ContractTerminationObserver;
+import lombok.val;
+import org.eclipse.edc.runtime.metamodel.annotation.Inject;
+import org.eclipse.edc.spi.event.EventEnvelope;
+import org.eclipse.edc.spi.event.EventRouter;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.system.ServiceExtension;
+import org.eclipse.edc.spi.system.ServiceExtensionContext;
+import org.eclipse.edc.spi.uuid.UuidGenerator;
+
+public class MdsLoggingHouseBinder implements ServiceExtension {
+
+    @Inject
+    private EventRouter eventRouter;
+
+    @Inject
+    private Monitor monitor;
+
+    @Inject
+    private ContractAgreementTerminationService contractAgreementTerminationService;
+
+    private ObjectMapper objectMapper;
+
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+
+        objectMapper = new ObjectMapper();
+        objectMapper.registerModule(new JavaTimeModule());
+
+        contractAgreementTerminationService.getContractTerminationObservable()
+            .registerListener(new ContractTerminationObserver() {
+                @Override
+                public void contractTerminationCompletedOnThisInstance(ContractTerminationEvent contractTerminationEvent) {
+                    val logEntry = LogEntry.from("contractTerminatedByThisInstance", contractTerminationEvent);
+
+                    sendMessage(contractTerminationEvent, logEntry);
+                }
+
+                @Override
+                public void contractTerminatedByCounterparty(ContractTerminationEvent contractTerminationEvent) {
+                    val logEntry = LogEntry.from("contractTerminatedByCounterparty", contractTerminationEvent);
+                    sendMessage(contractTerminationEvent, logEntry);
+                }
+            });
+    }
+
+    private void sendMessage(ContractTerminationEvent contractTerminationEvent, LogEntry logEntry) {
+        try {
+            val message = objectMapper.writeValueAsString(logEntry);
+            val event = new MdsContractTerminationEvent(
+                UuidGenerator.INSTANCE.generate().toString(),
+                contractTerminationEvent.contractAgreementId(),
+                message
+            );
+
+            @SuppressWarnings("unchecked")
+            EventEnvelope.Builder<MdsContractTerminationEvent> builder = EventEnvelope.Builder.newInstance();
+
+            val eventEnvelope = builder
+                .at(System.currentTimeMillis())
+                .payload(event)
+                .build();
+
+            eventRouter.publish(eventEnvelope);
+
+            monitor.debug("Published event for " + logEntry);
+        } catch (JsonProcessingException e) {
+            monitor.warning("Failed to serialize the event for the logging house " + logEntry);
+        }
+    }
```

Comment:
in general, this class does too much, but the least we can do is split these large functions into smaller functions, e.g.

```java

@Override
public void initialize(ServiceExtensionContext context) {
    objectMapper = new ObjectMapper();
    objectMapper.registerModule(new JavaTimeModule());

    contractAgreementTerminationService.getContractTerminationObservable()
        .registerListener(new ContractTerminationObserver() {
            @Override
            public void contractTerminationCompletedOnThisInstance(ContractTerminationEvent contractTerminationEvent) {
                onTerminationEvent(contractTerminationEvent, "contractTerminatedByThisInstance");
            }

            @Override
            public void contractTerminatedByCounterparty(ContractTerminationEvent contractTerminationEvent) {
                onTerminationEvent(contractTerminationEvent, "contractTerminatedByCounterparty");
            }
        });
}

@SneakyThrows
private void onTerminationEvent(ContractTerminationEvent event, String messageType) {
  val logEntry = LogEntry.from(messageType, contractTerminationEvent);
  val logEntryJson = objectMapper.writeValueAsString(logEntry);

  val event = new MdsContractTerminationEvent(
      UuidGenerator.INSTANCE.generate().toString(),
      contractTerminationEvent.contractAgreementId(),
      logEntryJson
  );

  publishEvent(event);
}

private void publishEvent(MdsContractTerminationEvent event) {
  @SuppressWarnings("unchecked")
  EventEnvelope.Builder<MdsContractTerminationEvent> builder = EventEnvelope.Builder.newInstance();

  val eventEnvelope = builder
      .at(System.currentTimeMillis())
      .payload(event)
      .build();

  eventRouter.publish(eventEnvelope);

  monitor.debug("Published event for " + logEntry);
}
```

Reply:
Agree for the class that has too many roles, I forgot that this is the Extension class. It's been extracted to `MdsContractTerminationObserver`.

I did a similar split but not the same because I want to handle log messages. I want a backup in case the info can't be sent to the logging house for whatever reason.